microservices/case-service/src/main/java/org/entando/bundle/service/impl/FileServiceImpl.java

```diff
@@ -0,0 +1,239 @@
+    }
+    return name;
+  }
+
+  private String getBuketUrl() {
+    String url = null;
+
+    try {
+      url = config.getS3().getBucket().getUrl();
+    } catch (Throwable t) {
+      log.error("Configuration error detected! No bucket url defined", t.getLocalizedMessage());
+    }
+    return url;
+  }
+
+  private S3Client getClient() {
+    // Create the S3Client object.
+    Region region = Region.EU_WEST_1;
+    S3Client s3 = S3Client.builder()
+      .credentialsProvider(EnvironmentVariableCredentialsProvider.create())
+      .region(region)
+      .build();
+    return s3;
+  }
+
+  @Override
+  public String fileUpload(File file, Map<String, String> metadata) {
+    S3Client s3 = getClient();
+
+    try {
+      PutObjectRequest putOb = PutObjectRequest.builder()
+        .bucket(getBuketName())
+        .key(file.getName())
+        .metadata(metadata)
+        .build();
+      PutObjectResponse response = s3.putObject(putOb, RequestBody.fromBytes(getObjectFile(file.getAbsolutePath())));
+      return response.eTag();
+    } catch (S3Exception t) {
+      log.error(t.awsErrorDetails().errorMessage());
+      t.printStackTrace();
+    }
+    return null;
+  }
+
+  private byte[] getObjectFile(String filePath) {
+    FileInputStream fileInputStream = null;
+    byte[] bytesArray = null;
+
+    try {
+      File file = new File(filePath);
+      bytesArray = new byte[(int) file.length()];
+      fileInputStream = new FileInputStream(file);
+      fileInputStream.read(bytesArray);
+
+    } catch (IOException e) {
+      log.error("error building byte array of file", e);
+    } finally {
+      if (fileInputStream != null) {
+        try {
+          fileInputStream.close();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+    return bytesArray;
+  }
+
+  @Override
+  public List<S3Object> fileList() {
+    S3Client s3 = getClient();
+
+    try {
+      ListObjectsRequest listObjects = ListObjectsRequest
+        .builder()
+        .bucket(getBuketName())
+        .build();
+
+      ListObjectsResponse res = s3.listObjects(listObjects);
+      log.debug("File list successfully downloaded from bucket");
+      return res.contents();
+    } catch (S3Exception t) {
+      log.error("Error getting the file list of a bucket", t.awsErrorDetails().errorMessage());
+    }
+    return null;
+  }
+
+  @Override
+  public File downloadFile(String keyName) {
+    S3Client s3 = getClient();
+    File dest = null;
+
+    try {
+      GetObjectRequest objectRequest = GetObjectRequest
+        .builder()
+        .key(keyName)
+        .bucket(getBuketName())
+        .build();
+
+      ResponseBytes<GetObjectResponse> objectBytes = s3.getObjectAsBytes(objectRequest);
+      byte[] data = objectBytes.asByteArray();
+
+      // Write the data to a local file.
+      dest = new File(FileHelper.getTmpPathForFile(keyName));
```

Comment:
simply use `File.createTempFile()`

Reply:
Didn't use that because I needed to preserve original filename (no randomic serial attached) for upload on S3